Update guide for testing (#2536)
PennyLaneAI/pennylane: doc/development/guide/tests.rst

Every test has to be added to the PennyLane `test folder <https://github.com/PennyLaneAI/pennylane/tree/master/tests>`__.
The test folder is following PennyLane structure, and therefore a test needs to be added to its corresponding folder.

Most tests do not require any interface (Autograd, Torch, TensorFlow and Jax). Tests without interface will be marked
as ``core`` tests automatically by pytest ``conftest.py``. For general tests, you can follow this example structure,
where it is only required to follow pytest guidelines:

.. code-block:: python

    import pennylane as qml

    def test_circuit_expval(self):
        """ Test that the circuit expectation value for PauliX is 0."""
        dev = qml.device("default.qubit", wires=1)

        @qml.qnode(dev)
        def circuit():
            return qml.expval(qml.PauliX(0))

        expval = circuit()

        assert expval == 0

This test will be marked automatically as a ``core`` test.

On the other hand, some tests require specific interfaces and need to marked in order to be run on our Github test suite.
Tests involving interfaces have to be marked with its respective marker: ``@pytest.mark.autograd``, ``@pytest.mark.torch``,

> ``@pytest.mark.tf`` and ``@pytest.mark.jax``. It is now prohibited to use  `ìmportorskip()`` inside tests with interfaces.

Reviewer: Minor question: what happens if you have a test that uses more than one interface?

Author: I used the marker ``pytest.mark.math`` for mixed interfaces, because the math module consists of mixed interfaces. It has a dedicated runner on GIthub where all interfaces are installed. I will detail that here.

Reviewer: Nice, and tests marked as such will be run by _all_ CI actions? I wonder if `pytest.mark.all` might help generalize it beyond the `math` module?

Author: No it has a dedicated runner called ``math`` and is only run once. But I agree that ``all`` could be a great marker.

if a developer was adding a new non-parametric operation, they could run:

.. code-block:: bash

    python -m pytest tests/ops/qubit/test_non_parametric_ops.py

Using ``pytest -m`` offers the possibility to select and run tests with specific markers. For example,
if Jax is installed and a developer wants to run only Jax related tests, they could run:

.. code-block:: bash

    python -m pytest tests -m "jax"

They exist markers for interfaces (``autograd``, ``torch``, ``tf``, ``jax``), for multiple interfaces (``math``) and
also for certain folders (``qchem`` and ``qcut``).

For running ``qchem`` tests, one can run the following:

.. code-block:: bash

    python -m pytest tests -m "qchem"

The slowest tests are marked with ``slow`` and can be deselected by:

.. code-block:: bash

    python -m pytest -m "not slow" tests

The ``pytest -m`` option support boolean combination of marker. It is therefore possible to run both Jax and TensorFlow
tests by writing:

.. code-block:: bash

    python -m pytest -m "jax and tf" tests

or Jax tests that are not slow:

.. code-block:: bash

>     python -m pytest -m "jax and not slow" tests

Reviewer: 😍

    pip install pytest pytest-cov pytest-mock flaky

Creating a test
~~~~~~~~~~~~~~~
Every test has to be added to the PennyLane `test folder <https://github.com/PennyLaneAI/pennylane/tree/master/tests>`__.
The test folder follows the structure of the PennyLane module folder. Therefore, tests needs to be added to the corresponding subfolder of the functionality they are testing.

> Most tests typcally will not require the use of an interface (such as Autograd, Torch, TensorFlow and Jax). Tests without an interface will be marked

Reviewer: ```suggestion
Most tests typcally will not require the use of an interface or autodifferentiation framework (such as Autograd, Torch, TensorFlow and Jax). Tests without an interface will be marked
```

as a ``core`` test automatically by pytest (the functionality for this is located in ``conftest.py``). For such general tests, you can follow the structure of the example below,

> where it is recommended that you follow general pytest guidelines:

Reviewer: Maybe good to turn this into a link to the pytest documentation?

Author: Yes added!

.. code-block:: python

    import pennylane as qml

    def test_circuit_expval(self):
        """ Test that the circuit expectation value for PauliX is 0."""
        dev = qml.device("default.qubit", wires=1)

        @qml.qnode(dev)
        def circuit():
            return qml.expval(qml.PauliX(0))

        expval = circuit()

        assert expval == 0

This test will be marked automatically as a ``core`` test.

On the other hand, some tests require specific interfaces and need to be marked in order to be run on our Github test suite.

> Tests involving interfaces have to be marked with their respective marker: ``@pytest.mark.autograd``, ``@pytest.mark.torch``,
> ``@pytest.mark.tf`` and ``@pytest.mark.jax``. If tests involve multiple interfaces, one should add the marker

Reviewer: To make this easier to read, I recommend having a bullet list here :)
```suggestion
Tests involving interfaces have to be marked with their respective marker:

- ``@pytest.mark.autograd``,

- ``@pytest.mark.torch``,

- ``@pytest.mark.tf``, and

- ``@pytest.mark.jax``.

If tests involve multiple interfaces, one should add the marker
```

``@pytest.mark.all_interfaces``. Additionally, it is now prohibited to use ``importorskip`` to import interface packages inside of a test. Interface tests
must be written separately for each interface and without the use of fixtures to span multiple interfaces. Instead, the package related
to interfaces can directly be imported in a marked test. Tests with interface markers are automatically skipped if the necessary

> interfaces are not installed.

Reviewer: Perhaps good to have this more prominently displayed? Like:

````rest
.. warning::

    Please do not use ``pytest.importorskip`` in your test files. Instead, simply
    import the autodifferentiation package as needed inside your marked test.
    The mark will automatically ensure that the test is skipped if the autodifferentiation
    framework is not installed.

````

> Tests that are not marked but do import an interface will lead to a failure in the Github test suite.

Reviewer: ```suggestion
Tests that are not marked but do import an interface will lead to a failure in the GitHub test suite.
```

    pip install pytest pytest-cov pytest-mock flaky

Creating a test
~~~~~~~~~~~~~~~
Every test has to be added to the PennyLane `test folder <https://github.com/PennyLaneAI/pennylane/tree/master/tests>`__.

> The test folder is following PennyLane structure, and therefore a test needs to be added to its corresponding folder.

Reviewer 2: ```suggestion
The test folder follows the structure of the PennyLane module folder. Therefore, tests needs to be added to the corresponding subfolder of the functionality they are testing.
```

> Most tests do not require any interface (Autograd, Torch, TensorFlow and Jax). Tests without interface will be marked
> as ``core`` tests automatically by pytest ``conftest.py``. For general tests, you can follow this example structure,
> where it is only required to follow pytest guidelines:

Reviewer 2: ```suggestion
Most tests typcally will not require the use of an interface (such as Autograd, Torch, TensorFlow and Jax). Tests without an interface will be marked
as a ``core`` test automatically by pytest (the functionality for this is located in ``conftest.py``). For such general tests, you can follow the structure of the example below,
where it is recommended that you follow general pytest guidelines:
```

.. code-block:: python

    import pennylane as qml

    def test_circuit_expval(self):
        """ Test that the circuit expectation value for PauliX is 0."""
        dev = qml.device("default.qubit", wires=1)

        @qml.qnode(dev)
        def circuit():
            return qml.expval(qml.PauliX(0))

        expval = circuit()

        assert expval == 0

This test will be marked automatically as a ``core`` test.

> On the other hand, some tests require specific interfaces and need to marked in order to be run on our Github test suite.
> Tests involving interfaces have to be marked with its respective marker: ``@pytest.mark.autograd``, ``@pytest.mark.torch``,
> ``@pytest.mark.tf`` and ``@pytest.mark.jax``. If tests involved multiple interfaces, one should add the marker
> ``@pytest.mark.all_interfaces``. It is now prohibited to use ``importorskip`` inside tests with interfaces. Also tests
> involving interfaces must be written separately and it should be avoided to use fixtures. The necessary packages related
> to interfaces have to be imported in the tests directly. All tests with marked interfaces are skipped if the necessary
> interfaces are not installed.

Reviewer 2: ```suggestion
On the other hand, some tests require specific interfaces and need to be marked in order to be run on our Github test suite.
Tests involving interfaces have to be marked with their respective marker: ``@pytest.mark.autograd``, ``@pytest.mark.torch``,
``@pytest.mark.tf`` and ``@pytest.mark.jax``. If tests involve multiple interfaces, one should add the marker
``@pytest.mark.all_interfaces``. Additionally, it is now prohibited to use ``importorskip`` to import interface packages inside of a test. Interface tests
must be written separately for each interface and without the use of fixtures to span multiple interfaces. Instead, the package related
to interfaces can directly be imported in a marked test. Tests with interface markers are automatically skipped if the necessary
interfaces are not installed.
```

> Tests that are not marked and also related to specific interfaces will be skipped in the Github test suite and therefore
> it will not be run and the results will not be added to the coverage (inducing a coverage fail).

Reviewer 2: Re our discussion, I think the missing marker here would lead to a failed test run, since the import statement should raise an error.
```suggestion
Tests that are not marked but do import an interface will lead to a failure in the Github test suite.
```

> You can find an example for testing a PennyLane template with Jax:

Reviewer 2: ```suggestion
Below you can find an example for testing a PennyLane template with Jax:
```

.. code-block:: python

    def circuit_template(features):
        qml.AngleEmbedding(features, range(3))
        return qml.expval(qml.PauliZ(0))

    def circuit_decomposed(features):
        qml.RX(features[0], wires=0)
        qml.RX(features[1], wires=1)
        qml.RX(features[2], wires=2)
        return qml.expval(qml.PauliZ(0))

    @pytest.mark.jax
    def test_jax(self, tol):
        """Tests the jax interface."""

        import jax
        import jax.numpy as jnp

        features = jnp.array([1.0, 1.0, 1.0])

        dev = qml.device("default.qubit", wires=3)

        circuit = qml.QNode(circuit_template, dev, interface="jax")
        circuit2 = qml.QNode(circuit_decomposed, dev, interface="jax")

        res = circuit(features)
        res2 = circuit2(features)
        assert qml.math.allclose(res, res2, atol=tol, rtol=0)

> You can also find an example of a test involving multiple interfaces:

Reviewer 2: ```suggestion
Another example of a test involving multiple interfaces is shown below:
```

            qml.AngleEmbedding(features, range(3))
            return qml.expval(qml.PauliZ(0))

        @pytest.mark.all_interfaces
        def test_all_interfaces_gradient_agree(self):
            """Test the results are similar between torch and tf"""
            import torch
            import tensorflow as tf

            dev = qml.device("default.qubit", wires=3)

            features_torch = torch.Tensor([1.0, 1.0, 1.0])
            features_tf = tf.Variable([1.0, 1.0, 1.0], dtype=tf.float64)

            circuit_torch = qml.QNode(circuit, dev, interface="torch")
            circuit_tf = qml.QNode(circuit, dev, interface="tf")

            res_torch = circuit_torch(features_torch)
            res_tf = circuit_tf(features_tf)

            assert np.allclose(res_torch, res_tf)

Running the tests
~~~~~~~~~~~~~~~~~

The `tests <https://github.com/PennyLaneAI/pennylane/tree/master/tests>`__ folder of the root PennyLane directory contains the PennyLane test suite. Run all tests in this folder via:

.. code-block:: bash

    python -m pytest tests

Using ``python -m`` ensures that the tests run with the correct Python version if multiple versions are on the system.

> As the entire test suite takes some time, locally running only relevant files speeds the debugging cycle. For example,

Reviewer 2: ```suggestion
As the entire test suite takes some time, locally running only relevant files speeds up the debugging cycle. For example,
```

if a developer was adding a new non-parametric operation, they could run:

.. code-block:: bash

    python -m pytest tests/ops/qubit/test_non_parametric_ops.py

Using ``pytest -m`` offers the possibility to select and run tests with specific markers. For example,
if Jax is installed and a developer wants to run only Jax related tests, they could run:

.. code-block:: bash

    python -m pytest tests -m "jax"

> They exist markers for interfaces (``autograd``, ``torch``, ``tf``, ``jax``), for multiple interfaces (``all_interfaces``) and
> also for certain folders (``qchem`` and ``qcut``).

Reviewer 2: ```suggestion
There exists markers for interfaces (``autograd``, ``torch``, ``tf``, ``jax``), for multiple interfaces (``all_interfaces``) and
also for certain PennyLane submodules (``qchem`` and ``qcut``).
```

For running ``qchem`` tests, one can run the following:

.. code-block:: bash

    python -m pytest tests -m "qchem"

The slowest tests are marked with ``slow`` and can be deselected by:

.. code-block:: bash

    python -m pytest -m "not slow" tests

> The ``pytest -m`` option support boolean combination of marker. It is therefore possible to run both Jax and TensorFlow
> tests by writing:

Reviewer 2: ```suggestion
The ``pytest -m`` option supports Boolean combinations of markers. It is therefore possible to run both Jax and TensorFlow
tests by writing:
```
